README.md
# Map Picker Flutter
Check it out at [Pub.Dev](https://pub.dev/packages/map_picker_flutter)

## This works on:
- [x] Android;
- [x] iOS;
- [x] Web;
- [x] Windows;
- [x] Linux;
- [x] MacOS;

![Screenshot_354](https://user-images.githubusercontent.com/22732544/111829863-009e1c80-88cc-11eb-9dc1-0e6dc3cfab59.png)

## Help Maintenance

I've been maintaining quite many repos these days and burning out slowly. If you could help me cheer up, buying me a cup of coffee will make my life really happy and get much energy out of it.

<a href="https://www.buymeacoffee.com/RtrHv1C" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/purple_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a>

## Objective
This lib was designed especially for web and desktop platforms, but of course you can also use it on mobiles.

## Getting Started
To use is simple, before all, you need to call this method, It will required once only,

    MapPicker.init(
            /// Custom search Builder Widget, case it is null, will be used the default of packer
            searchBuilder: (Function(String address) search) => MyCustomWidget(),

            /// Custom address Builder Widget, case it is null, will be used the default of packer
            addressBuilder: (String address, Function done) => MyCustomWidget(),

            /// Custom progress Widget, case it is null, will be used the default of packer
            progressWidget: MyCustomWidget(),

            /// Custom address Builder Widget, case it is null, will be used the default of packer
            marker: MyCustomWidget(),

            /// All above is to set global widgets to all times you need call this package

            theme: MPickerTheme(
                /// # Initial Location
                initialLocation: LatLng(-23.572143, -46.613275),

                /// # Text showed when address is not founded
                errorAddressMissing: 'Error, address don\'t founded.',

                /// # Text showed when gets error during find address
                errorToFindAddress: 'Error to find this address, try again.',

                /// # Initial text when don`t has any address selected
                withoutAddress: 'Without Address Picked',

                /// # Here the texts showed in [TextField] case you don`t use a custom
                searchHint: 'Type here...',
                searchLabel: 'Search:',

                /// # Language of return of google maps api
                lang: 'en'),

            /// My Google Maps Api Key
            key: 'GOOGLE_API_KEY');

# Then Usage

Now you can use of two ways

## Mobiles

You just call navigator and wait to response, ex:

    final MPAddress address = await Navigator.push(context, MaterialPageRoute(builder: (c) => MapPicker(
        /// here, case you want overwrite anything of global params of you set, you can

        /// # Custom search Builder Widget, case it is null, will be used the global or default
        searchBuilder: (Function(String address) search) => MyCustomWidget(),

        /// # Custom address Builder Widget, case it is null, will be used the global or default
        addressBuilder: (String address, Function done) => MyCustomWidget(),

        /// # Custom progress Widget, case it is null, will be used the global or default
        progressWidget: MyCustomWidget(),

        /// # Custom theme, case it is null, will be used the global or default
        theme: MPickerTheme(...),

        /// # Custom Marker, case it is null, will be used the global or default
        marker: MyWidget()
    )));

## Web / Desktop

In this platforms, because screens is bigger, I think is more intelligent use a dialog, then I made it ready for you to use

    final MPAddress address = await MapPickerTemplate.dialogAddressPicker(
        context: context);

Of course, you can use both ways on any platform, this is just a suggestion.

CHANGELOG.md
## [1.1.0+1] - Bug fix

## [1.1.0] - Null Safety

## [1.0.0+8] - Update dependencies

## [1.0.0+7] - Update dependencies

## [1.0.0+6] - Fix method return

## [1.0.0+5] - Fix method return

## [1.0.0+4] - Fix marker asset

## [1.0.0+3] - Fix marker

## [1.0.0+2] - Fix Pubspec

## [1.0.0+1] - Any fixes

## [1.0.0] - Add first release date.


lib/map_picker.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:map_picker_flutter/mpicker_address.dart';
import 'mpicker_controller.dart';
import 'package:map/map.dart';
import 'mpicker_theme.dart';

export 'package:latlng/latlng.dart';
export 'mpicker_templates.dart';
export 'mpicker_address.dart';
export 'mpicker_theme.dart';

import 'package:flutter/material.dart';

class MapPicker extends StatelessWidget {
  /// Global theme
  static late MPickerTheme _gTheme;

  /// Global searchBuilder Widget
  static Widget Function(Function(String address) search)? _searchBuilder;

  /// Global addressBuilder Widget
  static Widget Function(String address, Function done)? _addressBuilder;

  /// Global progressDialog Widget
  static Widget? _progressWidget;

  /// Global marker widget
  static Widget? _markerWidget;

  /// Google Maps API KEY
  static late String _key;

  /// # It need be called before you use this lib passing the theme and the Google Maps API KEY
  static init(
      {Widget Function(Function(String address) search)? searchBuilder,
      Widget Function(String address, Function done)? addressBuilder,
      Widget? progressWidget,
      Widget? marker,
      required MPickerTheme theme,
      required String key}) {
    MapPicker._addressBuilder = addressBuilder;
    MapPicker._progressWidget = progressWidget;
    MapPicker._searchBuilder = searchBuilder;
    MapPicker._markerWidget = marker;
    MapPicker._gTheme = theme;
    MapPicker._key = key;
  }

  /// Same of global variables but here case you want change in a specific called
  final Widget Function(Function(String address) search)? searchBuilder;
  final Widget Function(String address, Function done)? addressBuilder;
  final MPickerController controller;
  final Widget? progressWidget;
  final MPickerTheme theme;
  final Widget? marker;

  factory MapPicker({
    /// Custom search Builder Widget, case it is null, will be used the global or default
    Widget Function(Function(String address) search)? searchBuilder,

    /// Custom address Builder Widget, case it is null, will be used the global or default
    Widget Function(String address, Function done)? addressBuilder,

    /// Custom progress Widget, case it is null, will be used the global or default
    Widget? progressWidget,

    /// Custom theme, case it is null, will be used the global or default
    MPickerTheme? theme,

    /// Custom Marker, case it is null, will be used the global or default
    Widget? marker,
  }) =>
      MapPicker._internal(
          progressWidget:
              progressWidget ?? _progressWidget ?? CircularProgressIndicator(),
          addressBuilder: addressBuilder ?? _addressBuilder,
          searchBuilder: searchBuilder ?? _searchBuilder,
          marker: marker ?? _markerWidget,
          controller: MPickerController(
              theme: theme ?? _gTheme,
              key: _key,
              ctMap: MapController(
                  location: theme?.initialLocation ?? _gTheme.initialLocation,
                  zoom: theme?.zoom ?? _gTheme.zoom)),
          theme: theme ?? _gTheme);

  MapPicker._internal(
      {this.marker,
      this.searchBuilder,
      this.addressBuilder,
      this.progressWidget,
      required this.controller,
      required this.theme});

  dispose() => controller.dispose();

  Future pop(context) async {
    Navigator.pop(context, controller.popAddress);
    dispose();
  }

  @override
  Widget build(BuildContext context) {
    final width = MediaQuery.of(context).size.width;

    return Scaffold(
        body: Stack(
      children: [
        Center(
          child: Listener(
            onPointerSignal: controller.onPointerSignal,
            child: GestureDetector(
              onScaleEnd: (details) => controller.getAddressByLatLng(),
              onScaleUpdate: controller.onScaleUpdate,
              onScaleStart: controller.onScaleStart,
              onDoubleTap: controller.onDoubleTap,
              child: Map(
                controller: controller.ctMap,
                builder: (context, x, y, z) {
                  final url =
                      'https://mt1.google.com/vt/lyrs=m@129&x=$x&y=$y&z=$z';

                  return CachedNetworkImage(
                      errorWidget: (c, w, s) => CircularProgressIndicator(),
                      fit: BoxFit.contain,
                      imageUrl: url);
                },
              ),
            ),
          ),
        ),
        Center(
            child: marker ??
                Image.asset('assets/ic_location.png',
                    package: 'map_picker_flutter', height: 50, width: 50)),
        Positioned(
            right: 0,
            left: 0,
            top: 0,
            child: searchBuilder != null
                ? searchBuilder!(controller.getAddressByAddress)
                : Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Flexible(
                          child: Card(
                              shape: RoundedRectangleBorder(
                                  borderRadius: BorderRadius.circular(100)),
                              margin: EdgeInsets.all(20),
                              child: Container(
                                  margin: EdgeInsets.symmetric(horizontal: 15),
                                  width: width * .4 - 40,
                                  padding: EdgeInsets.only(
                                      bottom: 5, right: 15, left: 15),
                                  child: TextField(
                                      controller: controller.searchControl,
                                      decoration: InputDecoration(
                                          labelText: theme.searchLabel,
                                          hintText: theme.searchHint,
                                          prefixIcon: IconButton(
                                              icon: Icon(
                                                  Icons.arrow_back_outlined),
                                              onPressed: () =>
                                                  Navigator.pop(context)),
                                          suffixIcon: IconButton(
                                              icon: Icon(Icons.search),
                                              onPressed: () => controller
                                                  .getAddressByAddress(
                                                      controller.searchControl
                                                          .text))),
                                      onSubmitted:
                                          controller.getAddressByAddress)))),
                    ],
                  )),
        Positioned(
            bottom: 0,
            right: 0,
            left: 0,
            child: StreamBuilder<bool>(
                stream: controller.outStreamProgress,
                initialData: true,
                builder: (c, s) => (s.data ?? true)
                    ? (progressWidget ??
                        Row(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            Container(
                                margin: EdgeInsets.only(bottom: 15),
                                child: CircularProgressIndicator())
                          ],
                        ))
                    : StreamBuilder<MPAddress?>(
                        stream: controller.outStreamAddress,
                        builder: (c, ad) => (addressBuilder != null
                            ? addressBuilder!(
                                ad.data?.formattedAddress ??
                                    theme.withoutAddress,
                                dispose)
                            : Row(
                                mainAxisAlignment: MainAxisAlignment.center,
                                children: [
                                  Flexible(
                                      child: Card(
                                          shape: RoundedRectangleBorder(
                                              borderRadius:
                                                  BorderRadius.circular(10)),
                                          margin: EdgeInsets.all(15),
                                          child: Container(
                                              padding: EdgeInsets.symmetric(
                                                  horizontal: 25, vertical: 10),
                                              alignment: Alignment.center,
                                              width: width * .4 - 40,
                                              child: Text(
                                                  ad.data?.formattedAddress ??
                                                      theme.withoutAddress,
                                                  textAlign:
                                                      TextAlign.center))))
                                ],
                              )))))
      ],
    ));
  }
}


lib/mpicker_templates.dart
import 'package:flutter/material.dart';

export 'package:latlng/latlng.dart';
import 'map_picker.dart';
export 'map_picker.dart';

class MapPickerTemplate {
  /// # This feature is very useful in desktop or web, how the screen is bigger,
  /// the picker can be just a dialog, then it is done, done just to use.
  static Future<MPAddress> dialogAddressPicker(
      {Widget Function(Function(String address) search)? searchBuilder,
      Widget Function(String address, Function done)? addressBuilder,
      double aspectRatio = 2,
      Widget? progressWidget,
      MPickerTheme? theme,
      Widget? marker,
      required BuildContext context,
      String btnCancel = 'Cancel',
      String btnSave = 'Save'}) async {
    final content = MapPicker(
        progressWidget: progressWidget,
        addressBuilder: addressBuilder,
        searchBuilder: searchBuilder,
        marker: marker,
        theme: theme);

    return await showDialog(
        context: context,
        builder: (c) => Container(
            margin: EdgeInsets.only(bottom: 60),
            child: AlertDialog(
                content: AspectRatio(aspectRatio: aspectRatio, child: content),
                actions: [
                  TextButton(
                      child: Text(btnCancel),
                      onPressed: () {
                        Navigator.pop(context);
                        content.dispose();
                      }),
                  TextButton(
                      child: Text(btnSave),
                      onPressed: () async => await content.pop(context))
                ])));
  }
}


lib/mpicker_controller.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/gestures.dart';
import 'package:rxdart/rxdart.dart';
import 'package:latlng/latlng.dart';
import 'package:map/map.dart';
import 'package:dio/dio.dart';
import 'mpicker_address.dart';
import 'mpicker_theme.dart';
import 'dart:core' as c;
import 'dart:core';

class MPickerController with _Base {
  /// # Urls to get geocode infos
  static const _URL_ADDRESS =
      'https://maps.googleapis.com/maps/api/geocode/json?address=[ADDRESS]';
  static const _URL_LAT_LNG =
      'https://maps.googleapis.com/maps/api/geocode/json?latlng=[LAT],[LON]';

  final _progress = BehaviorSubject<bool>.seeded(false);
  final _address = BehaviorSubject<MPAddress?>();
  final MapController ctMap;
  final MPickerTheme theme;
  final String key;

  final searchControl = TextEditingController();

  Stream<MPAddress?> get outStreamAddress => _address.stream;
  Stream<bool> get outStreamProgress => _progress.stream;
  double _scaleStart = 1.0;
  c.bool _hasError = true;
  Offset? _dragStart;

  String get currentAddress => _address.value?.formattedAddress ?? '-';
  MPAddress? get popAddress => _hasError ? null : _address.value;

  MPickerController(
      {required this.key, required this.theme, required this.ctMap});

  /// Go To place with latLng
  goTo(LatLng? latLng) {
    if (latLng != null) {
      ctMap.center = latLng;
    }
  }

  /// Go To initial place again
  goToInitial() => goTo(theme.initialLocation);

  /// Zoom on double tap
  onDoubleTap() => ctMap.zoom += 0.5;

  /// On Star Change Scale
  onScaleStart(ScaleStartDetails details) {
    _dragStart = details.focalPoint;
    _scaleStart = 1.0;
  }

  /// On Change Scale
  onScaleUpdate(ScaleUpdateDetails details) {
    final scaleDiff = details.scale - _scaleStart;
    _scaleStart = details.scale;

    if (scaleDiff > 0) {
      ctMap.zoom += 0.02;
    } else if (scaleDiff < 0) {
      ctMap.zoom -= 0.02;
    } else {
      final now = details.focalPoint;
      final diff = now - (_dragStart ?? Offset(0, 0));
      _dragStart = now;
      ctMap.drag(diff.dx, diff.dy);
    }
  }

  /// Here is for web and desktop, when wheel scrolling of mouse, it will apply or remove zoom
  onPointerSignal(PointerSignalEvent e) {
    if (e is PointerScrollEvent) {
      final sl = e.scrollDelta.dy;

      if (sl > 0) {
        // to down
        if (ctMap.zoom > 3) {
          ctMap.zoom -= 0.5;
        }
      } else {
        // to up
        if (ctMap.zoom <= 20.5) {
          ctMap.zoom += 0.5;
        }
      }
    }
  }

  String _replaceIt(
          {required String text, required c.Map<String, String> remove}) =>
      remove.keys.fold<String>(text,
          (text, element) => text.replaceAll(element, remove[element] ?? ''));

  String _getExtraArgs() =>
      (theme.lang != null ? '&language=${theme.lang}' : '') + '&key=$key';

  /// # Get address by LatLnt
  getAddressByLatLng() async {
    final position = ctMap.center;
    _progress.sink.add(true);

    try {
      final resp = await Dio().get(_replaceIt(
          text: _URL_LAT_LNG + _getExtraArgs(),
          remove: {
            '[LAT]': position.latitude.toString(),
            '[LON]': position.longitude.toString()
          }));

      _address.sink.add(
          MPAddress(formattedAddress: _extractAddress(resp), latLng: position));
    } catch (msg) {
      _address.sink.add(MPAddress(
          formattedAddress: theme.errorToFindAddress, latLng: position));
      _hasError = true;
    }

    _progress.sink.add(false);
  }

  /// # Get address by text
  getAddressByAddress(String address) async {
    _progress.sink.add(true);

    try {
      final resp = await Dio().get(_replaceIt(
          text: _URL_ADDRESS + _getExtraArgs(),
          remove: {'[ADDRESS]': address.toString()}));

      goTo(_extractLatLng(resp));
      _address.sink.add(MPAddress(
          formattedAddress: _extractAddress(resp),
          latLng: _extractLatLng(resp)));
    } catch (msg) {
      _address.sink.add(MPAddress(formattedAddress: theme.errorToFindAddress));
      _hasError = true;
    }

    _progress.sink.add(false);
  }

  String _extractAddress(Response resp) {
    final emptyAddress = theme.errorAddressMissing;

    try {
      if (resp.data != null && resp.data['status'] == 'OK') {
        final address = resp.data['results'];

        if ((address as List).length > 0) {
          try {
            _hasError = false;
            return address[0]['formatted_address'].toString();
          } catch (msg) {
            _hasError = true;
            return emptyAddress;
          }
        }
      }

      _hasError = true;
      return emptyAddress;
    } catch (msg) {
      _hasError = true;
      return theme.errorToFindAddress;
    }
  }

  LatLng? _extractLatLng(Response resp) {
    try {
      if (resp.data != null && resp.data['status'] == 'OK') {
        final address = resp.data['results'];

        if ((address as List).length > 0) {
          final lg = address[0]['geometry']['location'];
          return LatLng(lg['lat'], lg['lng']);
        }
      }
    } catch (msg) {}

    return null;
  }

  @override
  void dispose() {
    _address.close();
    _progress.close();
  }
}

abstract class _Base {
  void dispose();
}


example/main.dart
import 'package:flutter/material.dart';
import './../lib/map_picker.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(title: 'Map Picker', home: MainScreen());
  }
}

class MainScreen extends StatelessWidget {
  final control = TextEditingController();

  @override
  Widget build(BuildContext context) {
    MapPicker.init(
        theme: MPickerTheme(

            /// # Initial Location
            initialLocation: LatLng(-23.572143, -46.613275),

            /// # Text showed when address is not founded
            errorAddressMissing: 'Error, address don\'t founded.',

            /// # Text showed when gets error during find address
            errorToFindAddress: 'Error to find this address, try again.',

            /// # Initial text when don`t has any address selected
            withoutAddress: 'Without Address Picked',

            /// # Here the texts showed in [TextField] case you don`t use a custom
            searchHint: 'Type here...',
            searchLabel: 'Search:',

            /// # Language of return of google maps api
            lang: 'pt-BR'),
        key: 'API_KEY_GOOGLE_MAPS');

    /// # Here a example to get address by a dialog
    return Scaffold(
        body: Center(
            child: ElevatedButton(
                onPressed: () async => print('address: ${await MapPickerTemplate.dialogAddressPicker(

                    /// # Here you can pass your custom widget to put text and search
                    searchBuilder: (search) => Card(child: TextField(decoration: InputDecoration(labelText: 'Search:', hintText: 'Type here...', prefixIcon: IconButton(icon: Icon(Icons.arrow_back_outlined), onPressed: () => Navigator.pop(context)), suffixIcon: IconButton(icon: Icon(Icons.search),

                        /// # To search, just call the method `search` passing the address string
                        onPressed: () => search(control.text))), controller: control)),

                    /// # Here your custom widget to show current selected address
                    addressBuilder: (txt, done) => Card(

                        /// # Case you want put a button here to done the process, you just need call `done` function
                        child: Container(padding: EdgeInsets.all(10), child: Text(txt, textAlign: TextAlign.center))),

                    /// # You can to pass your custom marker here
                    marker: Icon(Icons.location_pin, color: Colors.redAccent, size: 30),

                    /// # Text in negative button
                    btnCancel: 'Cancel',

                    /// # Text in positive button
                    btnSave: 'Save', context: context)}'),
                child: Text('Address Picker'))));
  }
}
